llvm/test/CodeGen/AMDGPU/extract-lowbits.ll

```diff
@@ -99,11 +99,21 @@ define i32 @bzhi32_b4_commutative(i32 %val, i32 %numlowbits) nounwind {
 ; ---------------------------------------------------------------------------- ;
 
 define i32 @bzhi32_c0(i32 %val, i32 %numlowbits) nounwind {
-; GCN-LABEL: bzhi32_c0:
-; GCN:       ; %bb.0:
-; GCN-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
-; GCN-NEXT:    v_bfe_u32 v0, v0, 0, v1
-; GCN-NEXT:    s_setpc_b64 s[30:31]
+; SI-LABEL: bzhi32_c0:
```

Comment:
This should be OK if numlowbits is known < bitwidth. This case is canonical, so you should bring back the predicate from the last pr

Reply:
Frankly speaking, I have difficulty in understanding what is canonical here.  Is the following IR still canonical?
  And do you mean we can check MaxActiveBits to allow the pattern?  Thanks.
    
  %low5bits      = and i32 %numlowbits, 31
  %numhighbits = sub i32 32, %numlowbits
  %mask = lshr i32 -1, %numhighbits
  %masked = and i32 %mask, %val

Reply:
run it through instcombine, and computeKnownBits as you did before

Reply:
> Run it through instcombine. Whatever it produces is canonical. and computeKnownBits as you did before
Running instcombine like this?
./opt -passes=instcombine canon.ll -S

Reply:
yes